orttraining/orttraining/python/training/ortmodule/experimental/hierarchical_ortmodule/_hierarchical_ortmodule.py

```diff
@@ -0,0 +1,194 @@
+                                do_constant_folding=False, export_params=False,
+                                keep_initializers_as_inputs=True,
+                                training=torch.onnx.TrainingMode.TRAINING)
+                    except Exception as e:
+                        exportable = False
+
+                exportable_list[module] = exportable
+                return exportable
+            else:
+                sub_exportable = True
+                for name, sub_module in sub_dict.items():
+                    if isinstance(sub_module, torch.nn.ModuleList):
+                        for name1, sub_module1 in sub_module._modules.items():
+                            sub_exportable1 = check_exportable(sub_module1)
+                            sub_exportable = sub_exportable and sub_exportable1
+                    else:
+                        sub_exportable1 = check_exportable(sub_module)
+                        sub_exportable = sub_exportable and sub_exportable1
+
+                if sub_exportable is False:
+                    # At least one existing sub-module is not exportable,
+                    # so is the entire module.
+                    exportable_list[module] = sub_exportable
+                    return sub_exportable
+                else:
+                    # Now, we know all sub-modules are exportable, so
+                    # we are going to check if the composition of them
+                    # is still exportable at this module level.
+                    module_exportable = True
+                    for args in module_arg_pool[module]:
+                        try:
+                            with tempfile.NamedTemporaryFile(prefix='sub-module') as temp:
+                                torch.onnx.export(
+                                    module, args, temp, opset_version=ONNX_OPSET_VERSION,
+                                    do_constant_folding=False, export_params=False,
+                                    keep_initializers_as_inputs=True,
+                                    training=torch.onnx.TrainingMode.TRAINING)
+                        except Exception as e:
+                            # If this module is not exportable for one arg
+                            # group, we say this module is not exportable.
+                            module_exportable = False
+                            # Already found a broken case.
+                            # No need to check next case.
+                            break
+
+                    exportable_list[module] = module_exportable
+                    return exportable_list[module]
+
+        # Add a hook to record forward's input for all modules.
+        recursive_hook(self._original_module)
+
+        # Run forward with actual input to record all possible
+        # inputs for all invoked modules.
+        y = self._original_module(*args, **kwargs)
```

Comment:
If this invocation adds value, may we replease y with `_`?
`_ = self._original_module(*args, **kwargs)`

Reply:
No problem.